chapter05/chapter05.adoc

```diff
@@ -0,0 +1,596 @@
+
+[source, java]
+----
+   @Inject
+   @ConfigProperty(name="product.maintenanceMode", defaultValue="false")
+   private boolean maintenanceMode;
+----
+
+In this example, the `product.maintenanceMode` property will be converted to an `Boolean` before it is injected into the `maintenanceMode` field. If the conversion fails, a `org.eclipse.microprofile.config.ConversionFailedException` will be thrown.
+
+=== Converting Configuration data to a POJO
+
+We can also use the Config API to convert our configuration data to a POJO:
+
+[source, java]
+----
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+
+public class MyApplication {
+    @Inject
+    private MaintenanceMessage message;
+}
+
+
+public class MaintenanceMessage {
+    @ConfigProperty(name="product.maintenanceMessage")
+    private String message;
+}
+----
+
+In this example, we’re injecting a property named "product.maintenanceMessage" into the message field of our MaintenanceMessage class. If the property does not exist, a `org.eclipse.microprofile.config.ConfigPropertyNotFoundException` will be thrown.
+
+=== Creating a Custom `ConfigSource`
+
+As we saw, the Config API makes it easy to inject configuration properties into an application. The Config API defines a contract for config implementations. A ConfigSource is used to read configuration data from a particular source. For example, we could create a ConfigSource that reads configuration data from a file.
+
+`ConfigSource` interface has the following methods: 
+- `String getName()` : Returns the name of the ConfigSource. 
+- `int getOrdinal()` : Returns the ordinal of the ConfigSource. Ordinals are used to determine the precedence of ConfigSources. A higher ordinal means a higher precedence. 
+- `Map<String, String> getProperties()` : Returns a map of the properties in this ConfigSource. The keys in the map are the property names, and the values are the property values. 
+- `getValue(String propertyName)` : Returns the value of the given property. If the property is not found, this method returns null. 
+- `Set getPropertyNames()` : Returns a Set of the property names in this ConfigSource.
+
+Let’s implement a feature in our MicroProfile e-Commerce application to integrate payment gateway configuration dynamically by creating a PaymentServiceConfigSource (a custom ConfigSource) which could fetch API keys and endpoints. This would ensure that payment service configurations are up-to-date and can be changed without redeploying the application.
+
+MicroProfile Config provides a flexible framework for integrating with various external configuration providers, enabling applications to load configurations from sources beyond the default system properties, environment variables, and microprofile-config.properties files. This capability is crucial for modern applications that may need to pull configuration from dynamic sources like cloud services, databases, or custom APIs.
```

Comment:
This paragraph is out of place. It should be placed in the beginning of this chapter.

Reply:
Fixed by 936f1ce31973eceb2c7eacf12deadb5ea3943b35 and 18da76cb09e9f2807fc28603c48d31b78bd468cf. The idea is to help readers understand and appreciate this capability with the help of the above code example.